### ISSUE TITLE
Why we "must train on the training inputs"?

### ISSUE THREAD

**Issue by KasperNiklas** [cornellius-gp/gpytorch#2008]
We are working with a non-stationary dataset with rapid, unpredictable variations described and shown in this earlier [Issue #1994](https://github.com/cornellius-gp/gpytorch/issues/1994#issue-1214405324). It is pretty tough to fit a gaussian process to such a problem with convergence errors, numerical instabilities and over-fitting. Especially the length scales are much smaller than what we deem would be fit. 

To combat the issue with over-fitting, we would like to evaluate the gradient on data which is not training data, such that the hyperparameters are fitted for a setting which is more representative of what it will be exposed to. We believed that this should be possible by giving `model()` some `test_x` data while it still is in `model.train()` mode. 

However, we encounter the error `You must train on the training inputs!` when trying that. 

This error is from the script [gpytorch/models/exact_gp.py: Line 243 - 258](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/models/exact_gp.py) and can be seen here.

```Python
 def __call__(self, *args, **kwargs):
        train_inputs = list(self.train_inputs) if self.train_inputs is not None else []
        inputs = [i.unsqueeze(-1) if i.ndimension() == 1 else i for i in args]

        # Training mode: optimizing
        if self.training:
            if self.train_inputs is None:
                raise RuntimeError(
                    "train_inputs, train_targets cannot be None in training mode. "
                    "Call .eval() for prior predictions, or call .set_train_data() to add training data."
                )
            if settings.debug.on():
                if not all(torch.equal(train_input, input) for train_input, input in zip(train_inputs, inputs)):
                    raise RuntimeError("You must train on the training inputs!")
            res = super().__call__(*inputs, **kwargs)
            return res
```

Our questions are now: 

- Why does `GPyTorch` enforce that only train data should be given in the forward pass during the train time? 
- Is there perhaps another way to circumvent this problem?

**Comment by GStechschulte** [comment]
I believe `GPyTorch` is expecting to _see_ the same data points for each training iteration. So if you pass in input data during training that it was not expecting, i.e, test data, then the error you are getting will be raised. Indeed, evaluating the model using test data, `test_x`, is a step in combating overfitting. 

One of the simplest techniques I first perform to address overfitting is to first split the data set into training and validation (test) sets and then:

Visualize the performance of the GP model on the validation set as a function of the training iterations (y-axs = accuracy or whatever metric you are using and x-axis = training iterations). This visualization can assist in identifying overfitting since optimization algorithms are iterative, they take many steps away from the initial parameter estimates. As training iterations increase, the accuracy on the training set will continue to increase whereas, at some point during the training iterations, accuracy on the validation set will decrease. Stopping training at that point of "divergence" is called **early stopping** and can be prevent the model from learning too much from the training data, i.e., overfitting. This method often works well in practice since we are restricting the ability of the optimization algorithm to transfer information from the training examples to the parameters.
